src/gui/userstatusdialogmodel.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) by Felix Weilbach <felix.weilbach@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#pragma once
+
+#include "common/result.h"
+
+#include <userstatusjob.h>
+#include <datetimeprovider.h>
+
+#include <QObject>
+#include <QMetaType>
+#include <QtNumeric>
+
+#include <cstddef>
+#include <memory>
+#include <vector>
+
+namespace OCC {
+
+class UserStatusDialogModel : public QObject
+{
+    Q_OBJECT
+
+    Q_PROPERTY(QString userStatusMessage READ userStatusMessage NOTIFY userStatusChanged)
+    Q_PROPERTY(QString userStatusEmoji READ userStatusEmoji WRITE setUserStatusEmoji NOTIFY userStatusChanged)
+    Q_PROPERTY(OCC::UserStatus::OnlineStatus onlineStatus READ onlineStatus WRITE setOnlineStatus NOTIFY onlineStatusChanged)
+    Q_PROPERTY(int predefinedStatusesCount READ predefinedStatusesCount NOTIFY predefinedStatusesChanged)
+    Q_PROPERTY(QStringList clearAtStages READ clearAtStages CONSTANT)
+    Q_PROPERTY(QString clearAt READ clearAt NOTIFY clearAtChanged)
+    Q_PROPERTY(QString errorMessage READ errorMessage NOTIFY errorMessageChanged)
+    Q_PROPERTY(QUrl onlineIcon READ onlineIcon CONSTANT)
+    Q_PROPERTY(QUrl awayIcon READ awayIcon CONSTANT)
+    Q_PROPERTY(QUrl dndIcon READ dndIcon CONSTANT)
+    Q_PROPERTY(QUrl invisibleIcon READ invisibleIcon CONSTANT)
+
+public:
+    explicit UserStatusDialogModel(QObject *parent = nullptr);
+
+    explicit UserStatusDialogModel(std::shared_ptr<UserStatusJob> userStatusJob,
+        QObject *parent = nullptr);
+
+    explicit UserStatusDialogModel(std::shared_ptr<UserStatusJob> userStatusJob,
+        std::unique_ptr<DateTimeProvider> dateTimeProvider,
+        QObject *parent = nullptr);
+
+    explicit UserStatusDialogModel(const UserStatus &userStatus,
+        std::unique_ptr<DateTimeProvider> dateTimeProvider,
+        QObject *parent = nullptr);
+
+    explicit UserStatusDialogModel(const UserStatus &userStatus,
+        QObject *parent = nullptr);
+
+    ~UserStatusDialogModel();
+
+    UserStatus::OnlineStatus onlineStatus() const;
+    Q_INVOKABLE void setOnlineStatus(OCC::UserStatus::OnlineStatus status);
+
+    QUrl onlineIcon() const;
```

Comment:
missing a `Q_REQUIRED_RESULT` before `QUrl`

same for all const methods returning a value

Reply:
Too bad we can't use C++17 [[nodiscard]]